### PR TITLE
ReportIssue: Use safe overloads for secondary locations

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Follow up of #9323 to prevent AD0001s with secondary locations raised on a referenced projects.

This PR:
* Replaces old usages with secondary locations with the new, safe API*
* Updates SE `ReportIssue` to the same principle

*Except one that will follow in a later PR

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs [hunk 1]
```diff
@@ -112,11 +112,11 @@ private void ReportIssue(IOperation operation, BasicBlock block, bool conditionV
         var secondaryLocations = SecondaryLocations(block, conditionValue, syntax.Span.End);
         if (secondaryLocations.Any())
         {
-            ReportIssue(Rule2583, syntax, secondaryLocations, properties: null, issueMessage, S2583MessageSuffix);
+            ReportIssue(Rule2583, syntax, secondaryLocations.Select(x => x.ToSecondary()), issueMessage, S2583MessageSuffix);
         }
         else
         {
-            ReportIssue(Rule2589, syntax, additionalLocations: null, properties: null, issueMessage, string.Empty);
+            ReportIssue(Rule2589, syntax, [], issueMessage, string.Empty);
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs:119]: This looks wrong. Can't we just remove the secondary location?
```suggestion
            ReportIssue(Rule2589, syntax, issueMessage, string.Empty);
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/RestrictDeserializedTypesBase.cs [hunk 2]
```diff
@@ -78,10 +78,8 @@ protected override ProgramState PreProcessSimple(SymbolicContext context)
         else if (UnsafeDeserialization(state, operation) is { } invocation)
         {
             var methodDeclaration = UnsafeMethodDeclaration(state, invocation.Instance);
-            var additionalLocations = methodDeclaration is not null
-                ? new[] { GetIdentifier(methodDeclaration).GetLocation() }
-                : Array.Empty<Location>();
-            ReportIssue(operation, additionalLocations, additionalLocations.ToProperties(SecondaryMessage), RestrictTypesMessage);
+            var additionalLocations = methodDeclaration is null ? [] : new[] { GetIdentifier(methodDeclaration).GetLocation().ToSecondary(SecondaryMessage) };
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/RestrictDeserializedTypesBase.cs:81]:
```suggestion
            var additionalLocations = methodDeclaration is null ? [] : [ GetIdentifier(methodDeclaration).GetLocation().ToSecondary(SecondaryMessage) ];
```

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM. Just two NITs.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs:119]: I tried, but it didn't work. We don't have overload with `Rule` parameter.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/RestrictDeserializedTypesBase.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/RestrictDeserializedTypesBase.cs:81]: I tried this one too, but it cannot infer the underlying type (array vs list vs enumerable vs... ). The VB version would work here, as that one always creates array.

**Pull request merged**